controller/sharding/sharding.go

```diff
@@ -131,6 +133,35 @@ func LegacyDistributionFunction(replicas int) DistributionFunction {
 	}
 }
 
+// LegacyDistributionByURLFunction is the same as LegacyDistributionFunction except
+// it will ensure that clusters with the same server URL will return the same shard
+func LegacyDistributionByURLFunction(replicas int) DistributionFunction {
+	return func(c *v1alpha1.Cluster) int {
+		if replicas == 0 {
+			log.Debugf("Replicas count is : %d, returning -1", replicas)
+			return -1
+		}
+		if c == nil {
+			log.Debug("In-cluster: returning 0")
+			return 0
+		}
+		if c.Shard != nil && int(*c.Shard) < replicas {
+			return int(*c.Shard)
+		}
+		server := c.Server
+		log.Debugf("Calculating cluster shard for cluster id=%s and server=%s", c.ID, server)
+		if server == "" {
+			return 0
+		} else {
+			h := fnv.New32a()
+			_, _ = h.Write([]byte(server))
+			shard := int32(h.Sum32() % uint32(replicas))
```

Comment:
Technically this can lead to overflows if replica count is between 2^31 and 2^32 (not that it'd happen).